modules/ROOT/pages/patterns/shortest-paths.adoc

```diff
@@ -0,0 +1,657 @@
+As each path pattern is looking for a single shortest path, there are two possible solutions.
+For the first of those solutions, the two paths follow different relationships:
+
+image::patterns_shortest_graph_pattern1.svg[width="400", role="middle"]
+
+For the second solution, however, both paths go via the link between `Droitwich Spa` and `Bromsgrove`:
+
+image::patterns_shortest_graph_pattern2.svg[width="400", role="middle"]
+
+The problem with this is that relationships are required to be unique within a given match to a graph pattern (for more information, see xref:patterns/reference.adoc#graph-patterns-rules-relationship-uniqueness[Syntax and semantics -> relationship uniqueness]).
+So if this second pair of paths were selected, then the query would return no results, in contrast to the first pair of paths.
+To avoid this non-determinism in the number of results, a rule is imposed whereby selective path patterns cannot be combined with any other path patterns in the same graph pattern.
+
+If it makes sense for there to be relationships appearing more than once, the path patterns can be solved by putting them in separate `MATCH` clauses:
+
+.Valid query
+[source,cypher]
+----
+MATCH (d:Station {name: 'Droitwich Spa'}),
+      (w:Station {name: 'Worcestershire Parkway Ll'}),
+      (c:Station {name: 'Cheltenham Spa'})
+MATCH p = SHORTEST 1 (d)--+(w)
+MATCH q = SHORTEST 1 (d)--+(c)
+RETURN [stop IN nodes(p) | stop.name] AS r1,
+       [stop IN nodes(q) | stop.name] AS r2
+----
+
+.Result
+[role="queryresult",options="header,footer",cols="2*m"]
+|===
+| r1 | r2
+
+| ["Droitwich Spa", "Bromsgrove", "Worcestershire Parkway Ll"]
+| ["Droitwich Spa", "Bromsgrove", "Cheltenham Spa"]
+
+2+d|Rows: 1
+|===
+
+
+== Planning shortest paths queries
+
+This section describes the operators used when planning shortest path queries.
+For readers not familiar with Cypher execution plans and operators, it is recommended to first read the section xref:planning-and-tuning/execution-plans.adoc[].
+
+There are two operators used to plan `SHORTEST` queries:
+
+* xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-all[`StatefulShortestPath(All)`] - uses a unidirectional breadth-first search algorithm to find shortest paths from a previously matched start node to an end node that has not yet been matched.
+
+* xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[`StatefulShortestPath(Into)`] - uses a bidirectional breadth-first search (BFS) algorithm, where two simultaneous BFS invocations are performed, one from the left boundary node and one from the right boundary node.
+
+
+`StatefulShortestPath(Into)` is used by the planner when both boundary nodes in the shortest path have been matched in a `MATCH` clause previous to that containing the `SHORTEST` keyword.
```

Comment:
The wording makes this seem to be a plan that is decided at runtime, rather than based on cardinality estimates. For example, on the UK railway dataset I get only one node matching each end node pattern, but it is still planned as `StatefulShortestPath(All)`:
```
PROFILE
MATCH (n:Station {name: "Denmark Hill"}), (m:Station WHERE m.name CONTAINS "Vic")
MATCH p = SHORTEST 1 (n)((l)-[e:LINK]-(r) WHERE e.distance < 10){,5}(m)
RETURN p
```

Reply:
For me a Cartesian Product is used to plan the first `MATCH` clause and then a `StatefulShortestPath(Into)` is used. @sherfert Could it be that a rewrite merges both `MATCH` clause together?

Reply:
On an empty database it does. But with the uk railway dataset it plans this:

```
Planner COST

Runtime PIPELINED

Runtime version 5.21

Batch size 128

+----------------------------+----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| Operator                   | Id | Details                                                | Estimated Rows | Rows | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Pipeline      |
+----------------------------+----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +ProduceResults            |  0 | p                                                      |              0 |    1 |      22 |              0 |                    7/0 |     2.486 |               |
| |                          +----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +Projection                |  1 | (n)-[e*]->(m) AS p                                     |              0 |    1 |       1 |                |                    1/0 |     1.949 |               |
| |                          +----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +StatefulShortestPath(All) |  2 | SHORTEST 1 (n) ((`l`)-[`e`]-(`r`)){0, 5} (m)           |              0 |    1 |    3107 |          89084 |                 1033/0 |    57.485 | In Pipeline 1 |
| |                          |    |         expanding from: n                              |                |      |         |                |                        |           |               |
| |                          |    |     inlined predicates: cache[e.distance] < $autoint_2 |                |      |         |                |                        |           |               |
| |                          |    |                         e:LINK                         |                |      |         |                |                        |           |               |
| |                          |    |                         m.name CONTAINS $autostring_1  |                |      |         |                |                        |           |               |
| |                          |    |                         m:Station                      |                |      |         |                |                        |           |               |
| |                          +----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +NodeIndexSeek             |  3 | RANGE INDEX n:Station(name) WHERE name = $autostring_0 |              1 |    1 |       2 |            376 |                    0/2 |     0.432 | In Pipeline 0 |
+----------------------------+----+--------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+

Total database accesses: 3132, total allocated memory: 89404
```
So it depends on an estimate (which in this case turns out to be wrong) at plan time. It can't know for sure how many will match the CONTAINS predicate. The current wording suggests it will find out after the MATCH is executed and choose the SSP(Into) operator then.

Reply:
Yes, I see. This clearly illustrates as issue with the current wording (and with my understanding of the code). 

When the query looks like this:
```
PROFILE
MATCH p = SHORTEST 1 (n:Station {name: "Denmark Hill"})((l)-[e:LINK]-(r) WHERE e.distance < 10){,5}(m:Station WHERE m.name CONTAINS "Vic")
RETURN p
```

Then it is based on estimates for the number of matches for `n` and `m`. Since the planner does not know that `m` will match with only one node, `StatefulShortestPath(Into)` will not be chosen.

However, for the query consisting of two `MATCH` clauses, my understanding was that the first `MATCH` clause would be planned before the second `MATCH` clause, connecting them using an `apply` operator. In that case, the second `MATCH` clause always gets as argument a single match for `n` and `m`. This would result is planning a `StatefulShortestPath(Into)` operator. This is clearly not happening. It looks like both `MATCH` clauses are merged together, which would make a lot of sense. I was only not aware of this step.